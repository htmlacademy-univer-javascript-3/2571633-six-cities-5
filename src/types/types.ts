import { store } from '../store';
import { AuthorizationStatus } from '../const';
export type City = {
  title: string;
  lat: number;
  lng: number;
};
export type OfferData = {

  offer: OfferIdDetails[];

  offerPageStatus: boolean;
};
export type OfferIdDetailsPage = {
  offer: OfferIdDetails;
  OfferIdDetailsPageStatus: boolean;

  nearbyOffers: OfferIdDetails[];

  comments: UserReview[];
};
export type OfferIdDetails = {
  id: string;
  title: string;
  type: string;
  price: number;
  city: {
    name: string;
    location: {
      latitude: number;
      longitude: number;
      zoom: number;
    };
  };
  location: {
    latitude: number;
    longitude: number;
    zoom: number;
  };
  isFavorite: boolean;
  isPremium: boolean;
  rating: number;
  description: string;
  bedrooms: number;
  goods: [string];
  host: {
    name: string;
    avatarUrl: string;
    isPro: boolean;
  };
  images: [string];
  maxAdults: number;

  previewImage: string;

};
export enum Cities {
  Paris = 'Paris',
  Cologne = 'Cologne',
  Brussels = 'Brussels',
  Amsterdam = 'Amsterdam',
  Hamburg = 'Hamburg',
  Dusseldorf = 'Dusseldorf',
}

export interface EmptyCityProps {
  city: Cities;
}

export type AuthorizationSlice = {
  authorizationStatus: AuthorizationStatus;
  userData: UserAuth | null;
  postError: boolean;
  userDataLoadingStatus: boolean;
  userEmail: string;
};
export type Point = {
  title: string;
  lat: number;
  lng: number;
};
export type OfferObject = {
  id: string;
  title: string;
  price: number;
  rating: number;
  type: string;
  isPremium: boolean;
  isFavorite: boolean;
  NumberOfPlaces: number;
  previewImage: string;
  city: {
    name: string;
    location: {
      latitude: number;
      longitude: number;
      zoom: number;
    };
  };
  location: {
    latitude: number;
    longitude: number;
    zoom: number;
  };
};
export enum AppRoute {
  Login = '/login',
  Favorites = '/favorites',
  Main = '/',
  Offer = '/offer/:id',
}

export const enum CardCssNameList {
  citiesList = 'cities__card',
  neardPlace = 'near-places__card',
  favoritePlace = 'favorites__card',
}

export type UserReview = {
  id: number;
  user: {
    name: string;
    avatar: string;
  };
  rating: number;
  comment: string;
  date: string;
};
export type UserReviewPost = {
  rating: number;
  comment: string;
  id: string;
};
export type UserObject = {
  name: string;
  avatarUrl: string;
  isPro: boolean;
};
export const enum SortName {
  popular = 'popular',
  lowToHigh = 'lowToHigh',
  highToLow = 'highToLow',
  topRated = 'topRated',
}

export type UserAuth = UserObject & {
  email: string;
  token: string;
};

export type LoginAuth = {
  email: string;
  password: string;
};
export type ReducerType = {
  offer: OfferIdDetailsPage;
  user: UserAuth;
  review: UserReview;
};
export type Points = Point[];
export type State = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
