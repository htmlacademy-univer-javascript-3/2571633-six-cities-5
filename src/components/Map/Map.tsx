import {useRef, useEffect} from 'react';
import leaflet, {Icon} from 'leaflet';
import useMap from '../../hooks/use-map';
import { OfferIdDetails, City } from '../../types/types';

import 'leaflet/dist/leaflet.css';
const defaultCustomIcon = new Icon({
  iconUrl: 'https://assets.htmlacademy.ru/content/intensive/javascript-1/demo/interactive-map/pin.svg',
  iconSize: [40, 40],
  iconAnchor: [20, 40]
});

const currentCustomIcon = new Icon({
  iconUrl: 'https://assets.htmlacademy.ru/content/intensive/javascript-1/demo/interactive-map/main-pin.svg',
  iconSize: [40, 40],
  iconAnchor: [20, 40]
});

const activeCustomIcon = new Icon({
  iconUrl: '../../markup/img/pin-active.svg',
  iconSize: [40, 40],
  iconAnchor: [20, 40]
});
type MainPageProps = {

  offers: OfferIdDetails[] | undefined;
  currentCity: City;
  selectedPoint: OfferIdDetails | undefined;

  activeOffer: string | null;
};

function Map(props: MainPageProps): JSX.Element {
  const {offers, currentCity,activeOffer,selectedPoint} = props;

  const mapRef = useRef(null);
  const map = useMap(mapRef, currentCity);

  useEffect(() => {
    if (map) {
      map.eachLayer((layer) => {
        if (layer instanceof leaflet.Marker) {
          layer.remove();
        }
      });

      offers?.forEach((offer) => {
        leaflet
          .marker({
            lat: offer.location.latitude,
            lng: offer.location.longitude,
          }, {
            // eslint-disable-next-line no-nested-ternary
            icon: activeOffer === offer.id ? activeCustomIcon : selectedPoint !== undefined && offer.title === selectedPoint.title ? currentCustomIcon : defaultCustomIcon,
          })
          .addTo(map);
      });
      map.flyTo({lat: currentCity.lat,lng:currentCity.lng}, 12);
    }

  }, [map, offers, activeOffer, selectedPoint, currentCity.lat, currentCity.lng]);


  return (<div style={{ height: '100%' }} ref={mapRef}></div>);
}

export default Map;
