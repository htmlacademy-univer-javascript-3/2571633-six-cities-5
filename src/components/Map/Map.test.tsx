import { render } from '@testing-library/react';

import Map from './Map';
import {OfferIdDetails } from '../../types/types.ts';

it('Map correctly renders', () => {
  const mockCity = {
    name: 'Paris',
    location: { latitude: 48.8566, longitude: 2.3522, zoom: 13 },
  };

  const mockPlaces: OfferIdDetails[] = [
    {
      id: '1',
      title: 'Place 1',
      type: 'apartment',
      price: 100,
      location: { latitude: 48.857, longitude: 2.354, zoom: 10 },
      isFavorite: false,
      isPremium: true,
      rating: 4.5,
      city: mockCity,
      description: '',
      bedrooms: 0,
      goods: [''],
      host: {
        name: '',
        avatarUrl: '',
        isPro: false
      },
      images: [''],
      maxAdults: 0,
      previewImage: ''
    },
    {
      id: '2',
      title: 'Place 2',
      type: 'house',
      price: 200,
      location: { latitude: 48.858, longitude: 2.355, zoom: 10 },
      isFavorite: true,
      isPremium: false,
      rating: 4.0,
      city: mockCity,
      description: '',
      bedrooms: 0,
      goods: [''],
      host: {
        name: '',
        avatarUrl: '',
        isPro: false
      },
      images: [''],
      maxAdults: 0,
      previewImage: ''
    },
  ];

  const tree = render(<Map offers={mockPlaces} currentCity={{ title: 'Paris',lat: 48.8566, lng: 2.3522}} selectedPoint={undefined} activeOffer={null}/>);

  expect(tree).toMatchSnapshot();
});
