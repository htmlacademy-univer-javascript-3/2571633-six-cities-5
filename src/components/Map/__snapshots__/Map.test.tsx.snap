// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Map correctly renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
        data-testid="map-test"
        style="height: 100%; position: relative;"
        tabindex="0"
      >
        <div
          class="leaflet-pane leaflet-map-pane"
          style="left: 0px; top: 0px;"
        >
          <div
            class="leaflet-pane leaflet-tile-pane"
          >
            <div
              class="leaflet-layer "
              style="z-index: 1;"
            >
              <div
                class="leaflet-tile-container leaflet-zoom-animated"
                style="z-index: 18; left: 0px; top: 0px;"
              >
                <img
                  alt=""
                  class="leaflet-tile"
                  role="presentation"
                  src="https://a.basemaps.cartocdn.com/rastertiles/voyager/12/2074/1409.png"
                  style="width: 256px; height: 256px; left: -195px; top: -36px;"
                />
              </div>
            </div>
          </div>
          <div
            class="leaflet-pane leaflet-shadow-pane"
          />
          <div
            class="leaflet-pane leaflet-overlay-pane"
          />
          <div
            class="leaflet-pane leaflet-marker-pane"
          >
            <img
              alt=""
              class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
              src="https://assets.htmlacademy.ru/content/intensive/javascript-1/demo/interactive-map/pin.svg"
              style="margin-left: -20px; margin-top: -40px; width: 40px; height: 40px; left: 6px; top: -2px; z-index: -2; z-index: -2;"
              tabindex="0"
            />
            <img
              alt=""
              class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
              src="https://assets.htmlacademy.ru/content/intensive/javascript-1/demo/interactive-map/pin.svg"
              style="margin-left: -20px; margin-top: -40px; width: 40px; height: 40px; left: 8px; top: -6px; z-index: -6;"
              tabindex="0"
            />
          </div>
          <div
            class="leaflet-pane leaflet-tooltip-pane"
          />
          <div
            class="leaflet-pane leaflet-popup-pane"
          />
        </div>
        <div
          class="leaflet-control-container"
        >
          <div
            class="leaflet-top leaflet-left"
          >
            <div
              class="leaflet-control-zoom leaflet-bar leaflet-control"
            >
              <a
                aria-label="Zoom in"
                class="leaflet-control-zoom-in"
                href="#"
                role="button"
                title="Zoom in"
              >
                +
              </a>
              <a
                aria-label="Zoom out"
                class="leaflet-control-zoom-out"
                href="#"
                role="button"
                title="Zoom out"
              >
                −
              </a>
            </div>
          </div>
          <div
            class="leaflet-top leaflet-right"
          />
          <div
            class="leaflet-bottom leaflet-left"
          />
          <div
            class="leaflet-bottom leaflet-right"
          >
            <div
              class="leaflet-control-attribution leaflet-control"
            >
              <a
                href="https://leafletjs.com"
                title="A JS library for interactive maps"
              >
                Leaflet
              </a>
               | © 
              <a
                href="https://www.openstreetmap.org/copyright"
              >
                OpenStreetMap
              </a>
               contributors © 
              <a
                href="https://carto.com/attributions"
              >
                CARTO
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
      data-testid="map-test"
      style="height: 100%; position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="https://a.basemaps.cartocdn.com/rastertiles/voyager/12/2074/1409.png"
                style="width: 256px; height: 256px; left: -195px; top: -36px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-pane leaflet-shadow-pane"
        />
        <div
          class="leaflet-pane leaflet-overlay-pane"
        />
        <div
          class="leaflet-pane leaflet-marker-pane"
        >
          <img
            alt=""
            class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
            src="https://assets.htmlacademy.ru/content/intensive/javascript-1/demo/interactive-map/pin.svg"
            style="margin-left: -20px; margin-top: -40px; width: 40px; height: 40px; left: 6px; top: -2px; z-index: -2; z-index: -2;"
            tabindex="0"
          />
          <img
            alt=""
            class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
            src="https://assets.htmlacademy.ru/content/intensive/javascript-1/demo/interactive-map/pin.svg"
            style="margin-left: -20px; margin-top: -40px; width: 40px; height: 40px; left: 8px; top: -6px; z-index: -6;"
            tabindex="0"
          />
        </div>
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        >
          <div
            class="leaflet-control-zoom leaflet-bar leaflet-control"
          >
            <a
              aria-label="Zoom in"
              class="leaflet-control-zoom-in"
              href="#"
              role="button"
              title="Zoom in"
            >
              +
            </a>
            <a
              aria-label="Zoom out"
              class="leaflet-control-zoom-out"
              href="#"
              role="button"
              title="Zoom out"
            >
              −
            </a>
          </div>
        </div>
        <div
          class="leaflet-top leaflet-right"
        />
        <div
          class="leaflet-bottom leaflet-left"
        />
        <div
          class="leaflet-bottom leaflet-right"
        >
          <div
            class="leaflet-control-attribution leaflet-control"
          >
            <a
              href="https://leafletjs.com"
              title="A JS library for interactive maps"
            >
              Leaflet
            </a>
             | © 
            <a
              href="https://www.openstreetmap.org/copyright"
            >
              OpenStreetMap
            </a>
             contributors © 
            <a
              href="https://carto.com/attributions"
            >
              CARTO
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
